### PR TITLE
[CI][E2E] Split gen12 pre-commit testing into build and run stages

### PR DESCRIPTION
**Pull request by ayylol**
This patch enables E2E build and run splitting on pre-commit CI. Currently one of the main bottlenecks in pre-commit CI is the Linux Gen12 testing which takes between 30 ~ 40 minutes to complete. By instead building these tests on our fast build systems, and then transferring the binaries to the Gen12 machine we can cut down on this time by roughly half.

Sadly, the Windows Gen12 testing will still be the main bottleneck so total time for pre-commit won't improve with this PR, but we will at least get faster feedback for gen12 on Linux.

### CODE REVIEW

File: sycl/test-e2e/format.py
Comment (ayylol) [inline sycl/test-e2e/format.py:1]: Changes here are to work-around a hang that occurs when one of our tests try to run a binary that does not exist when inside one of the containers that we use for CI. 

This hang is not a recent regression but rather an issue that has existed for a while but is very infrequently encountered with the current full testing mode. With the split testing if a test fails to build a binary in the build-only stage it will fail the run-only stage due to the binary not existing, making this hang a lot more common.

File: .github/workflows/sycl-linux-precommit.yml
Comment (aelovikov-intel) [inline .github/workflows/sycl-linux-precommit.yml:76]:
```suggestion
  build_e2e_tests:
```

File: .github/workflows/sycl-linux-precommit.yml
Comment (aelovikov-intel) [inline .github/workflows/sycl-linux-precommit.yml:93]:
```suggestion
  run_prebuilt_e2e_tests:
```

File: .github/workflows/sycl-linux-precommit.yml
Comment (aelovikov-intel) [inline .github/workflows/sycl-linux-precommit.yml:91]: Don't remember why we needed this. Maybe because builds are release and we wanted some widely used format there. If that's the case, it's different for e2e tests and we can just hard-code the one that performs better.

File: .github/workflows/sycl-linux-precommit.yml
Comment (aelovikov-intel) [inline .github/workflows/sycl-linux-precommit.yml:85]: Is it possible to skip this?

File: .github/workflows/sycl-linux-run-tests.yml
Comment (aelovikov-intel) [inline .github/workflows/sycl-linux-run-tests.yml:195]: Is this used anywhere? What's the purpose of it?

File: .github/workflows/sycl-linux-run-tests.yml
Comment (aelovikov-intel) [inline .github/workflows/sycl-linux-run-tests.yml:334]: Why do we need second part of the expression?

File: .github/workflows/sycl-linux-run-tests.yml
Comment (aelovikov-intel) [inline .github/workflows/sycl-linux-run-tests.yml:364]: That should be handled by lit, e.g. `--param sycl_devices="${{ inputs.target_devices }}"`

File: .github/workflows/sycl-linux-run-tests.yml
Comment (aelovikov-intel) [inline .github/workflows/sycl-linux-run-tests.yml:452]: Why `inputs.upload_artifact`? Can't we use build-only mode as the condition?

Also, how much space is this going to require (mostly interested in comparison with the toolchain artifact itself).

File: sycl/test-e2e/format.py
Comment (aelovikov-intel) [inline sycl/test-e2e/format.py:296]: Another option is to only run pre-built binaries if that build task was 100% successful. Not sure if we want that, probably depends on the CI load patterns after the changes land and everything stabilizes.

Comment (ayylol) [reply sycl/test-e2e/format.py:296]: The reason i did not do this initially is because we can have a situation where the build fails on a test that would not be ran on a particular runner, so I think its still useful to run the testing.

Also for example if an unrelated change makes a single test on the build fail you would still want to see if your changes make any other tests fail at the run stage.

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml:195]: This is not used anywhere, I added it because I was mirroring the way `sycl-linux-build.yml` handles outputting its artifact.

We could use it to cancel the running workflows if the build fails, but I wasnt a big fan of that idea so I did not implement that.

I can remove this if that's preferable.

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml:452]: The size of the artifact is 270MB right now.

right now build-only/run-only modes are set through the extra_lit_opts input, which may have other options included in it. so I needed another variable to tell me if I should upload the binaries. 
What I could do instead is make a parameter which we would set to build-only/run-only/full and use that in these checks, as well as use that to add the lit option for the testing mode. Open to other ideas as well.

File: sycl/test-e2e/format.py
Comment (aelovikov-intel) [reply sycl/test-e2e/format.py:296]: That's one side of a coin. Another is that in 99% cases the author will upload an updated patch fixing those e2e-build issues and it's not unlikely that e2e-run-only results will be "wasted". If we can ensure that e2e-build failure is reported fast enough, then not running e2e in this scenario might improve the overall CI responsiveness.

File: .github/workflows/sycl-linux-run-tests.yml
Comment (aelovikov-intel) [reply .github/workflows/sycl-linux-run-tests.yml:195]: Up to @sarnex , @uditagarwal97 . My preference is not to introduce dead code until it becomes necessary.

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml:334]: I'm not too certain, I was basing this bit off of the `Download SYCL toolchain` step. Looks like that was added in this commit 936dac7965abec530588aef1e73a1b66a9f1a6fd.

File: .github/workflows/sycl-linux-precommit.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-precommit.yml:91]: We also do this in the existing e2e pre-commit testing 
https://github.com/intel/llvm/blob/sycl/.github/workflows/sycl-linux-precommit.yml#L145-L147

File: .github/workflows/sycl-linux-precommit.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-precommit.yml:85]: In what scenarios should we skip this? Im guessing you are suggesting adding an if statement for cases in which we dont want to run this? im not too clear on this.

Comment (aelovikov-intel) [reply .github/workflows/sycl-linux-precommit.yml:85]: Sorry for not being clear, I meant skip/remove that line.

Comment (sarnex) [reply .github/workflows/sycl-linux-precommit.yml:85]: yeah unclear to me why we need to set target devices as well

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [reply .github/workflows/sycl-linux-run-tests.yml:195]: agree

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [inline .github/workflows/sycl-linux-run-tests.yml:460]: i dont see `retention-days` as input, so i dont understand how this can not error out

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [reply .github/workflows/sycl-linux-run-tests.yml:452]: i think if we set the retention days to like 1 day or something it shouldnt be that bad?

File: sycl/test-e2e/format.py
Comment (sarnex) [reply sycl/test-e2e/format.py:1]: can we at least make a GH issue for the hang with repro steps?

File: sycl/test-e2e/format.py
Comment (sarnex) [reply sycl/test-e2e/format.py:296]: imo i like the CI still running as much as possible even if something fails

File: .github/workflows/sycl-linux-precommit.yml
Comment (sarnex) [reply .github/workflows/sycl-linux-precommit.yml:91]: that reasoning sounds somewhat familiar

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [inline .github/workflows/sycl-linux-run-tests.yml:105]: could be im just dense but its not clear to me what possible values of this be from the var name. maybe we can rename it?

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [inline .github/workflows/sycl-linux-run-tests.yml:209]: i wonder if this thing will work

https://github.com/ihiroky/extract-action

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [inline .github/workflows/sycl-linux-run-tests.yml:343]: is this just to save disk space?

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [inline .github/workflows/sycl-linux-run-tests.yml:338]: i dont really see any setup going on, so i think the Extract name is fine

File: sycl/test-e2e/AOT/cpu.cpp
Comment (AlexeySachkov) [inline sycl/test-e2e/AOT/cpu.cpp:10]: What is the reason behind this extra requirement?

If we have any limitations now as to what we can do in our tests, then they should be documented somewhere.
Also, will we still continue to have `build-and-run` configuration somewhere in our pre-commit? I'm just double-checking that we are not disabling those tests forever.

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml:456]: Not sure either, I'll add it and set it to 1

File: sycl/test-e2e/format.py
Comment (ayylol) [reply sycl/test-e2e/format.py:296]: I think another thing to consider in the scenario I highlighted is what we currently do when that occurs. Usually we'll point to a ticket and comment that the failure is unrelated and ask gatekeepers to merge as everything else passes. However if we don't run the testing we won't have this information, and effectively block the progress in such PRs.

Comment (aelovikov-intel) [reply sycl/test-e2e/format.py:296]: I think the "flaky" fails are happening 99.9% of the time on execution and not when building, so it might not be that bad. We'd just have to be quicker with the reverts when something does break. That said, if @sarnex wants to run as much as possible, then do that.

File: sycl/test-e2e/format.py
Comment (ayylol) [reply sycl/test-e2e/format.py:1]: Yep, here it is: #16351

File: .github/workflows/sycl-linux-run-tests.yml
Comment (KornevNikita) [reply .github/workflows/sycl-linux-run-tests.yml:210]: @sarnex just wondering - is there any policy regarding regarding usage of non-github actions?

File: sycl/test-e2e/AOT/cpu.cpp
Comment (ayylol) [reply sycl/test-e2e/AOT/cpu.cpp:10]: Hey, so the extra requirements in this pr fell into these camps:
- The tests inside the `Adapter` folder were testing sycl-ls, the `RUN:` lines there don't use `%{run}` or `%{run-unfiltered-devices}` so they were being ran on the build system. We could wrap these lines in `%if run-mode` but i figured it would just be cleaner to mark them as requires `build-and-run`, since they have no lines that compile to begin with
- The majority of the other tests were AOT cpu tests that fail on the run-side. These fail because when compiling the AOT tests we target the host machine's cpu architecture, which may be different than that of the running machine. So to get these tests running on separate machines we would need to restructure them to use the `-Xsycl-target-backend --march=` and then on the running side make sure we are executing the appropriate binary based on the cpu's architecture.
- the tests in `OptionalKernelFeatures` and `SpecConstants` that were changed failed on the build side due to member functions not being found. I'm honestly not too sure why/what the fix should be for these.

Here's a run of the CI before adding the extra requirements: https://github.com/intel/llvm/actions/runs/12262397938

As for the comment about the tests being disabled forever, I am working on #16306 to avoid this. The idea there being that when in `run-only` mode if we have a test marked as `REQUIRES: build-and-run` we know we will not have the binary available from the `build-only` step, so we should fully run the test, including any compilation steps.

Should also point out Andrei's comment on that pr https://github.com/intel/llvm/pull/16306#discussion_r1881197021, Which I think explains the current plan in regards to this extra mark-up.

File: sycl/test-e2e/Adapters/sycl-ls-gpu-default-any.cpp
Comment (ayylol) [reply sycl/test-e2e/Adapters/sycl-ls-gpu-default-any.cpp:2]: Not currently, I can make a PR for changes test-e2e readme adding those details.

For now here's the extra considerations I've come across with the split testing:
- In tests that build multiple binaries and execute them we need to make sure that each binary has a different name so that they are not overwritten, and we can transfer all the binaries to the run system.
- For XFAILS we have two scenarios to consider now. 
  - If the expected failure occurs during compilation we need to mark the XFAIL on a feature that is device-agnostic (i.e., `XFAIL: linux` or `XFAIL: *`). If in this situation we mark a test as XFAIL with a device-specific feature we will get a normal failure, since the `build-only` mode will not have any device-specific features.
  - If the expected failure occurs during run-time we will need to mark the test with XFAIL on a device specific feature (i.e., `XFAIL: gpu`, `XFAIL: level_zero`), or if its expected to always fail on run-time we can use `XFAIL: run-mode`. If in this situation we mark a test as `XFAIL` with a device-agnostic feature we will get an unexpected pass on the build stage, since the test will be able to compile.
- To do the test splitting we classify `RUN:` lines as either build or run. if a line contains `%{run}`, `%{run-unfiltered-devices}` or `%if run-mode` they are classified as run-lines, otherwise they are classified as build-lines. 
  - Previously we could get away with having lines that executed the test binary without `%{run}` or `%{run-unfiltered-devices}`, however now this will usually create a failure at the build stage as we will try to execute the binary without the appropriate devices.
  - The vast majority of `RUN:` lines that do not use `%{run}`/`%{run-unfiltered-devices}` are needed to compile the binary, or need to be ran prior to the execution (but not necessarily on the machine doing the execution). in the case that we need to run a command after executing the test binaries, and using either of the run expansions does not make sense, we can use `%if run-mode` to make sure it runs on the running system instead of the build system.
- Right now the build-only mode does not support logic to properly assess the features in REQUIRES/UNSUPPORTED to know if a test should be built in the environment (or for `spir64`), Only tests that are marked `REQUIRES: build-and-run-mode` or `UNSUPPORTED: true` are skipped. so if a test will fail building for the build environment we have on CI or for spir64 we will need to mark this as `REQUIRES: build-and-run-mode`. This is only temporary, and I am working on a way to not need these types of requires statements.
- Other than that there are the AOT tests, and tests which do not compile anything at all. Which I mentioned in my response to Alexey above.

Comment (kbenzie) [reply sycl/test-e2e/Adapters/sycl-ls-gpu-default-any.cpp:2]: That's a good chunk of info :+1: to documenting this in the README.

**Review by kbenzie — APPROVED** [review]
UR LGTM

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml:344]: I'm not certain, but I think so. I did this to mirror the `Extract/Setup SYCL toolchain` step that we already have. I can remove it if that would be preferable.

File: .github/workflows/sycl-linux-precommit.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-precommit.yml:89]: Hard-coded the archive name as well as the compression/decompression commands.

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml]: Ok, i removed it and tried it on the standalone mode here https://github.com/intel/llvm/actions/runs/12361572315/job/34499021175 It didn't try to run this step so i think its fine.

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml]: I ended up hardcoding the type of compression we use, so I dont think this is necessary anymore

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml]: Removed this and it seems to work fine still

File: sycl/test-e2e/Adapters/sycl-ls-gpu-default-any.cpp
Comment (ayylol) [reply sycl/test-e2e/Adapters/sycl-ls-gpu-default-any.cpp:2]: @kbenzie @AlexeySachkov Made #16409 to add info in the README

File: .github/workflows/sycl-linux-run-tests.yml
Comment (aelovikov-intel) [inline .github/workflows/sycl-linux-run-tests.yml:324]: Does it work when `target_devices` is unset?

**Review by aelovikov-intel — APPROVED** [review]

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [inline .github/workflows/sycl-linux-run-tests.yml:69]: can we add a `description` describing all the options and what each means?

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [inline .github/workflows/sycl-linux-run-tests.yml:411]:
```suggestion
    - name: Pack E2E binaries
```

File: .github/workflows/sycl-linux-run-tests.yml
Comment (sarnex) [inline .github/workflows/sycl-linux-run-tests.yml:414]:
```suggestion
    - name: Upload E2E binaries
```

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml:324]: Yep, It's set to `all` by default if this parameter is empty, and on build-only we change that to an empty list.

File: sycl/test-e2e/OneapiDeviceSelector/illegal_input.cpp
Comment (sarnex) [inline sycl/test-e2e/OneapiDeviceSelector/illegal_input.cpp:3]: sorry what does `--crash` do?

**Review by sarnex — COMMENTED** [review]
lgtm, minor comments!

really nice work, gen12 testing is the bane of my existence cutting in half is incredible!

File: .github/workflows/sycl-linux-run-tests.yml
Comment (ayylol) [reply .github/workflows/sycl-linux-run-tests.yml:76]: Ok, added

File: sycl/test-e2e/OneapiDeviceSelector/illegal_input.cpp
Comment (ayylol) [reply sycl/test-e2e/OneapiDeviceSelector/illegal_input.cpp:3]: The workaround for the lit hang that I was experiencing involved calling the `_runShTest` function with the `useExternalSh` parameter set to true (previously this was false). When lit uses the internal shell it has some special logic for handling the `not` command which I think made it unnecessary to add the `--crash` flag here. However this failed after adding the workaround because the underlying executable that was called crashed instead of exiting normally with a non-zero exit code, and the `not` command in the external shell needed this flag to properly react to that crash.

I also ran this test using the internal shell + the `--crash` flag and it works fine too.

Comment (sarnex) [reply sycl/test-e2e/OneapiDeviceSelector/illegal_input.cpp:3]: ah ok thanks for the explanation, i thought it was running bash `not` here and i couldn't find any info about that arg, but we're actually calling an llvm utility also named `not`, where the `--crash` argument makes it return true if the thing crashes. the more you know.

**Review by sarnex — APPROVED** [review]
lgtm! thx

**Review by YuriPlyakhin — APPROVED** [review]
Joint Matrix changes LGTM

File: sycl/test-e2e/AOT/cpu.cpp
Comment (ayylol) [reply sycl/test-e2e/AOT/cpu.cpp:10]: @AlexeySachkov Hey, just wanted to check if your concerns with the documentation + accidentally disabling tests were addressed. With the documentation pr #16409, and the use of the fallback option.

With the regards to the fallback option for running tests that require `build-and-run`, I've confirmed that on the split run-only side we are now running the tests marked with `REQUIRES: build-and-run-mode` that are able to run on gen12. (For example the `KernelAndProgram/test_cache_jit_aot.cpp` test which is marked this way ran, and passed there)

**Comment by ayylol** [comment]
@intel/dpcpp-sanitizers-review friendly ping

**Review by AllanZyne — APPROVED** [review]

**Comment by ayylol** [comment]
@sarnex FYI, I had to change the container for the build-only stage from the ubuntu 22 alldeps container to the ubuntu 24 latest intel drivers container because some tests started to fail on it. I was using that container originally because it has the dependencies from the build container so that we can build the cuda/hip tests as well, but since those arent enabled yet for this pr its not really needed yet.

**Comment by sarnex** [comment]
Ok yeah looks like the problem is some tests link against `libze_loader` which is part of the GPU driver, and they recently dropped support for 22.04 (or at least don't provide packages for it) so I switched us to using 24.04, so `libze_loader` is using a newer glibc so it won't work on 24.04. Eventually we should make a 24.04 alldeps image but for now this is fine.

**Comment by ayylol** [comment]
@intel/llvm-gatekeepers This is ready to merge.
Failures in post commit gen12 for tests `OnlineCompiler/online_compiler_OpenCL.cpp` and `USM/fill_any_size.cpp`, are unrelated to this pr and appearing currently in post-commit on sycl branch (https://github.com/intel/llvm/actions/runs/12437990459/job/34729407798)

**Comment by sarnex** [comment]
@ayylol Can we/are we planning to do the same for postcommit?

**Comment by ayylol** [comment]
> @ayylol Also, seeing an XPASS during building the E2E tests, I don't know what an XPASS even means there but yeah
> 
> ```
> Unexpectedly Passed Tests (1):
>   SYCL :: Matrix/SG32/get_coordinate_ops.cpp
> ```
> 
> https://github.com/intel/llvm/actions/runs/12438967580/job/34732118174

XPASS on the build-only stage usually means that its a test that is able to compile, but fails when running but the way it is marked the `XFAIL` is incorrectly triggered on the build-only. Opened #16455 to fix this.

> @ayylol Can we/are we planning to do the same for postcommit?

Yep, adding this to the post-commit should be quite similar, likely wont require all the extra test markup I had to do in this pr. I'm thinking we should hold on it and see if the pre-commit changes aren't causing too much hassle before proceeding with that.

**Comment by sarnex** [comment]
> Yep, adding this to the post-commit should be quite similar, likely wont require all the extra test markup I had to do in this pr. I'm thinking we should hold on it and see if the pre-commit changes aren't causing too much hassle before proceeding with that.

Sure, sounds good, thx!

**Comment by aelovikov-intel** [comment]
IMO, post-commit should happen after

1) We moved all pre-commit (except PVC e2e maybe)
2) We got back and cleaned up the logic for unsupported/xfail/etc so that it works nicely with the feature and doesn't require too much explicit markup. 

To clarify: "REQURIES: pvc, build-and-run-mode" is bad, "REQUIRES-RUN: pvc" or "XFAIL-RUN: pvc" is much better. I'm not pushing for these specific spellings, but the whole "build-and-run-mode" was a temporary solution to make fast progress and rip the immediate benefits in the pre-commit.

**Pull request merged**
